### ISSUE TITLE
Chapter 4 - Testing section - 'get user by slug failed ECONNREFUSED'

### ISSUE THREAD

**Issue by andrewshrout** [async-labs/saas#120]
**Whats happening?**
Page doesn't load. Get this error. Consoles in VScode are fine.
`request to http://localhost:8000/api/v1/public/get-user-by-slug failed, reason: connect ECONNREFUSED 127.0.0.1:8000
`

**What am I expecting?**
To see the index.ts render and see the email and name from the database.

**What am I doing?**
I up both yarn dev's. I get the same thing as the tutorial in the console, using the same code. I double checked and traced all my environment variables that they're in the right place pointing at the right ports, commented out the code that isn't even referenced / exists yet (all the ./team stuff), and added stuff to the next.config.js even though the tutorial doesn't say to, but you get linting errors.

I keep checking various stack overflows on this error but none of it makes sense to me. It looks like each path its interacting with has error catches, but this doesn't give any information and it hangs way too long.

API ENV
```
PORT_API=8000
URL_APP=http://localhost:3000
URL_API=http://localhost:8000
MONGO_URL=myurl
```
APP ENV
```
URL_APP=http://localhost:3000
URL_API=http://localhost:8000
PORT_APP=3000
```

**To Recreate**
https://github.com/andrewshrout/saas-book-course/tree/master/book/4-begin

Yarn dev in both app and api.

<p/><hr/><details>
      <summary>Click to see <b>Hill for issue #120</b></summary>
      <div><img src="https://async-github-hills.s3.amazonaws.com/async-labs-saas/120.png" alt="Single issue hill" class="s3-image" width="100%" /></div></details>
      <details><summary>Click to see <b>Hill for all issues </b></summary><div>
      <img src="https://async-github-hills.s3.amazonaws.com/async-labs-saas/all-issue-hill.png" alt="All issue hill" class="s3-image" width="100%" /></div></details>
      
created by [Async](https://async-await.com/)

**Comment by tima101** [comment]
@andrewshrout I am not on my working laptop so I cannot reproduce. But try starting `API` first and then starting `APP`. Then load `Index` page. If same problem persists, I will check code tomorrow.

PS: Please use `Chapter 4` instead of `Book 4` in title. It may confuse other readers. We only have two books. Thank you.

**Comment by andrewshrout** [comment]
Problem persists. I fire up API, and receive the ready. I fire up app, it bundles successfully and I get the ready. When I go to localhost:3000 in the browser, I get
```
[ event ] build page: /
[ wait ]  compiling ...
[ info ]  bundled successfully, waiting for typecheck results...
[ ready ] compiled successfully
```

but the page remains hanging. The API console tells me its firing Express Route and Static Method. Then eventually I get the econn failure on the api link.

**Comment by tima101** [comment]
@andrewshrout I can reproduce that page does not load in:

https://github.com/andrewshrout/saas-book-course/tree/master/book/4-begin

I looked at code `API` server and I think page does not load because a few things missing:
- CORS between `API` and `APP`
- `API` connection to MongoDB database (server calls `User.getUserBySlug` inside Express route `/get-user-by-slug`)

Do we ask you, reader, to test both apps at this codebase in the book?
If so, please send me some sentence so I can find this place and fix it.
It is too early for testing, codebase lacks the above two pieces.

**Comment by andrewshrout** [comment]
@tima101 ch4 -> "Testing Connection" section. The entire section of it up until you get to Jest is about this. (it also prevents you from doing any of the slugify testing, as your pages don't render)

**Comment by tima101** [comment]
@andrewshrout Looks like codebase you shared with me did not code that connects to database. Would please add it to https://github.com/andrewshrout/saas-book-course/tree/master/book/4-begin and let me try again?

**Comment by andrewshrout** [comment]
Sorry. I think I forgot to push last night. @tima101 Just fixed that.

**Comment by tima101** [comment]
@andrewshrout You need to connect `API` server to MongoDB in the this file.  https://github.com/andrewshrout/saas-book-course/blob/master/book/4-begin/api/server/server.ts Also books suggests creating first MongoDB document manually.

**Comment by andrewshrout** [comment]
I must've deleted that code at some point when I was inspecting why "log" wasn't working. I did create the data manually.

When I fix that issue, I get `TypeError: Cannot read property 'email' of undefined` on render. Adding console.log()s to my post method in api/server/api/public.ts shows me that I'm getting my userdoc, and it contains an email.

`{
  _id: 5f5a824cefe8a807dc2164bd,
  email: 'team@builderbook.org',
  displayName: 'Team Builder Book'
}`

Which means my response isn't coming through back to App right? I trace the path of the functions, and double checked my env. It should be getting it. Using console.logs() on the appside in sendRequest shows me that the data makes it app side, and is returned from the tryblock as data.

So the issue is in 
```
export const getUserBySlugApiMethod = (slug) =>
  sendRequestAndGetResponse(`${BASE_PATH}/get-user-by-slug`, {
    body: JSON.stringify({ slug }),
  });
```

or in getInitialProps? 

I'm trying to figure this out so I can solve these issues on my own but its somewhat frustrating.

[repushed]

**Comment by tima101** [comment]
@andrewshrout Alright, that's a progress. I guess we need to worry about CORS yet at this point.
```
TypeError: Cannot read property 'email' of undefined
```

Replace:
```
    const userDoc = await User.getUserBySlug({ slug });
    console.log(userDoc)
    res.json({ userDoc });
```

With:
```
    const user = await User.getUserBySlug({ slug });
    console.log(user)
    res.json({ user });
```

user object was nested by one level. Fixed in the content. Thanks!

**Comment by andrewshrout** [comment]
Works. @tima101 Thanks for the quick responses as always. Sorry to pester so much as I work through this!

**Comment by tima101** [comment]
@andrewshrout As long as you don't change code from the book's - it's all good. Thanks for paying attention to details, this helps us too.
